homeassistant/components/binary_sensor/ads.py

```diff
@@ -0,0 +1,78 @@
+"""Support for ADS binary sensors."""
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import BinarySensorDevice, \
+    PLATFORM_SCHEMA, DEVICE_CLASSES_SCHEMA
+from homeassistant.components.ads import DATA_ADS, CONF_ADSVAR
+from homeassistant.const import CONF_NAME, CONF_DEVICE_CLASS
+import homeassistant.helpers.config_validation as cv
+
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['ads']
+DEFAULT_NAME = 'ADS binary sensor'
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ADSVAR): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Binary Sensor platform for ADS."""
+    ads_hub = hass.data.get(DATA_ADS)
+    if not ads_hub:
+        return False
+
+    adsvar = config.get(CONF_ADSVAR)
+    name = config.get(CONF_NAME)
+    device_class = config.get(CONF_DEVICE_CLASS)
+
+    ads_sensor = AdsBinarySensor(ads_hub, name, adsvar, device_class)
+    add_devices([ads_sensor], True)
+
+    ads_hub.add_device_notification(adsvar, ads_hub.PLCTYPE_BOOL,
```

Comment:
To avoid errors in the entity callback. the registration of the callback should move to a method on the entity called `async_added_to_hass` which is a coroutine that will be called when the entity has been added to home assistant. See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/android_ip_webcam.py#L259

Reply:
I tried to implement it. Not sure if it's right. Please take a look.